### ISSUE TITLE
Fix incorrect discount code expiration date format (discount codes go inactive when they shouldn't)

### ISSUE THREAD

**Issue by mintplugins** [easydigitaldownloads/easy-digital-downloads#5619]
Discount codes have been going inactive when they shouldn't. This is because the way that discount code expiration dates are saved to the database makes it difficult to properly query them using date comparison. This causes incorrect dates to be returned by the cron that does the discount code auto-expiration. It's why this issue was a thing:

https://github.com/easydigitaldownloads/easy-digital-downloads/issues/4838

Right now the expiration dates are saved into a format with the month first:
```
'm/d/Y H:i:s'
```

In order to do a proper query using `<`, the year needs to be first, month second, and day 3rd. We'll have to do an upgrade routine which changes all of the discount code expiration date formats to put the year first, month second, and day third. 

See where that is being done here:
https://github.com/easydigitaldownloads/easy-digital-downloads/blob/904db487f6c07a3a46903202d31d4e8ea2b30808/includes/discount-functions.php#L1012

See this for background: http://stackoverflow.com/questions/3847736/comparing-two-dates

**Comment by leewillis77** [comment]
Just my 2p , but I'd suggest unix timestamps rather than yyyymmdd format.

**Comment by pippinsplugins** [comment]
@leewillis77 that's the plan but we have to do a complete migration routine before we can do that and we have to account for backwards compatibility.

**Comment by cklosowski** [comment]
With the new EDD Discounts table, I believe we've solved this. Closing it as putting in the efforts now just seems like we're doing twice the work after #5277

**Comment by cklosowski** [comment]
@sunnyratilal if you disagree, please re-open.

**Comment by stracker-phil** [comment]
In case you want to provide a quick-fix for the current DB format: You can compare the current date format via SQL. 

I've added a custom con task on my site that successfully deletes expired discount codes using the MySQL function "[`STR_TO_DATE`](https://www.w3schools.com/sql/func_mysql_str_to_date.asp)" (available since MySQL 4.0):

```php
<?php
// Prepare SQL to fetch the 25 oldest expired discount codes.
$sql = $wpdb->prepare(
	"SELECT post.ID
	FROM {$wpdb->posts} AS post
	INNER JOIN {$wpdb->postmeta} AS meta_date ON post.ID = meta_date.post_id
	WHERE
		post.post_type = 'edd_discount'
		AND meta_date.meta_key = '_edd_discount_expiration'
		AND LENGTH(meta_date.meta_value) > 0
		AND STR_TO_DATE(meta_date.meta_value, '%%m/%%d/%%Y') < %s // ← that's the important condition
	ORDER BY post.ID ASC
	LIMIT 25;",
	current_time( 'Y-m-d' )
);

// Get the post IDs of those discount coupons.
$coupon_ids = $wpdb->get_col( $sql );
```
